test/core/event_engine/test_suite/tests/dns_test.cc

```diff
@@ -25,7 +63,529 @@ void InitDNSTests() {}
+constexpr char kHealthCheckRecordName[] =
+    "health-check-local-dns-server-is-alive.resolver-tests.grpctestingexp";
+
+// You know where this comes from.
+template <typename T>
+void WaitForSingleOwner(std::shared_ptr<T>&& object) {
+  while (object.use_count() > 1) {
+    GRPC_LOG_EVERY_N_SEC(2, GPR_INFO, "engine.use_count() = %ld",
+                         object.use_count());
+    absl::SleepFor(absl::Milliseconds(100));
+  }
+}
+
+}  // namespace
+
+class EventEngineDNSTest : public EventEngineTest {
+ public:
+  static void SetUpTestSuite() {
+    // Invoke bazel's executeable links to the .sh and .py scripts (don't use
+    // the .sh and .py suffixes) to make sure that we're using bazel's test
+    // environment.
+    std::string kPythonWrapper = "tools/distrib/python_wrapper";
+    std::string kDNSServerPath = "test/cpp/naming/utils/dns_server";
+    std::string kDNSResolverPath = "test/cpp/naming/utils/dns_resolver";
+    std::string kTCPConnectPath = "test/cpp/naming/utils/tcp_connect";
+    // HACK: Hyrum's law.
+    if (!grpc_core::GetEnv("TEST_SRCDIR").has_value()) {
+      // Invoke the .sh and .py scripts directly where they are in source code.
+      kPythonWrapper += ".sh";
+      kDNSServerPath += ".py";
+      kDNSResolverPath += ".py";
+      kTCPConnectPath += ".py";
+    }
+    // 1. launch dns_server
+    int port = grpc_pick_unused_port_or_die();
+    ASSERT_NE(port, 0)
+        << "pick unused port failed, maybe the port server is not running? "
+           "Start it with tools/run_tests/start_port_server.py";
+    // <path to python wrapper> <path to dns_server.py> -p <port> -r <path to
+    // records config>
+    _dns_server.server_process = new grpc::SubProcess(
+        {kPythonWrapper, kDNSServerPath, "-p", std::to_string(port), "-r",
+         kDNSTestRecordGroupsYamlPath});
+    // TODO(yijiem): no way to check whether it fails or not
+    _dns_server.port = port;
+
+    // 2. wait until dns_server is up (health check)
+    bool health_check_succeed = false;
+    for (int i = 0; i < 10; i++) {
+      // 2.1 tcp connect succeeds
+      // <path to python wrapper> <path to tcp_connect.py> -s <hostname> -p
+      // <port>
+      grpc::SubProcess tcp_connect({kPythonWrapper, kTCPConnectPath, "-s",
+                                    "localhost", "-p", std::to_string(port)});
+      // TODO(yijiem): no way to check whether it fails or not
+      if (tcp_connect.Join()) {
+        // 2.2 make an A-record query to dns_server
+        // <path to python wrapper> <path to dns_resolver.py> -s <hostname> -p
+        // <port> -n <domain name to query>
+        std::string command = absl::StrJoin(
+            {kPythonWrapper.c_str(), kDNSResolverPath.c_str(), "-s",
+             "127.0.0.1", "-p", std::to_string(port).c_str(), "-n",
+             kHealthCheckRecordName},
+            " ");
+        // TODO(yijiem): build a portable solution for Windows
+        FILE* f = popen(command.c_str(), "r");
```

Comment:
Another good reason to consider doing this in another tool. Python might make this scripting a lot cleaner.

Reply:
Can we use the same infrastructure as resolver_component_test here?  It would be better to avoid reinventing any wheels here.

Reply:
(This comment becomes outdated after the latest push)

1. The `dns_test` launches a local python dns_server, uses another python script to do the health checking and uses the server to answer queries and runs tests against the `EventEngine::DNSResolver` implementation. Part of the challenges is to handle runfiles in both bazel and blaze build systems. The way which the `resolver_component_test` handles this part really feels like tech-debt to me, e.g. https://source.corp.google.com/piper///depot/google3/third_party/grpc/bazel/grpc_build_system.bzl;l=1593-1602?q=file:third_party%2Fgrpc%20alternative&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental

So in this change I added a minimal bazel version of `test/cpp/util/get_grpc_test_runfile_dir.(h|cc)` and a minimal blaze version of `third_party/grpc/google_specific/test/cpp/util/get_grpc_test_runfile_dir.cc` in the imported CL: https://critique.corp.google.com/cl/543840102/depot/google3/third_party/grpc/google_specific/test/cpp/util/get_grpc_test_runfile_dir.cc and switching between them in `grpc_build_system.bzl` (again feel like tech-debt, but maybe less...).

2. The `resolver_component_test` test infra is not really flexible to be reused. That said, I added a `test/cpp/naming/utils/health_check.py` script and move the health checking logic from the `dns_test` into this script. This script could potentially be reused as well.

Reply:
I feel like we might still be duplicating some stuff unnecessarily here, but I might be wrong, and I don't have time to really dig in to understand the details.  I'm going to defer to @apolcyn for review of the tests here, with particular focus on making sure that we are not needlessly reinventing any wheels.